src/ui.jsx

```diff
@@ -0,0 +1,392 @@
+        );
+
+    }
+
+}
+
+/**
+ * The header bar at the top of the page.
+ */
+class Header extends React.Component {
+
+    render() {
+        return (
+            <div id="header">
+                <p>Martian Robots</p>
+            </div>
+        );
+    }
+
+}
+
+/**
+ * State entirely resides here; if it's need elsewhere, it can be passed down to child components as props.
+ */
+class World extends React.Component {
+
+    constructor() {
+
+        super();
+
+        /**
+         * State becomes immutable with some easily updatable properties. Every tick of the world updates the entire
+         * world state so that React can re-render it.
+         * @type {Object}
+         */
+        this.state = {
+
+            // Every robot that exists in the world and its state
+            robots: [],
+
+            // The state of the planet
+            planet: {
+                scents: {
+                    // E.g. "5,2": true
+                },
+                rows: 0,
+                cols: 0
+            },
+
+            outputMessages: []
+
+        };
+
+        this.setIntervalId = -1;
+
+        // TODO: Used for debugging, push state to this array after every tick so we can examine it later
+        this.savedStates = [];
+
+    }
+
+    /**
+     * Set the initial state of the world.
+     * @param  {String} instructionsString The user's input taken straight from the editor.
+     */
+    _setup(instructionsString) {
+
+        var instructionStack;
+
+        // Convert the instructions passed in by the user into a stack
+        try {
+            instructionStack = parseInstructions(instructionsString);
+        } catch (error) {
+            console.log(error);
+            return;
+        }
+
+        // The first item on the stack is the planet's boundaries
+        var planetBoundaries = instructionStack.pop().trim().split(" ");
+
+        // Use the instruction stack to set the initial state of the world
+        var initialRobots = prepareRobots(instructionStack, planetBoundaries);
+        var initialPlanet = { scents: {}, rows: parseInt(planetBoundaries[0], 10), cols: parseInt(planetBoundaries[1], 10) };
+
+        // Update the state of the world accordingly: The callback starts the simulation
+        this.setState({ robots: initialRobots, planet: initialPlanet },
+
+            () => {
+                this.setIntervalId = setInterval(this._tick.bind(this), 100);
+            }
+
+        );
+
+    }
+
+    // Update everything in the world by one instruction
+    // TODO: Make all of the data accesses immutable
+    _tick() {
+
+        // Use executeInstruction once on each robot to get the next state of the world
+        var planet = this.state.planet;
+        var robots = this.state.robots;
+        var outputMessages = this.state.outputMessages.slice(0);
+        var doneCount = 0;
+
+        // Map is immutable, return result to new array
+        var newRobots = robots.map( robot => {
```

Comment:
Both `map/forEach` should always pass objects by reference. If you want to prevent this, freeze the object. (immutable data structures will fix this more elegantly when you get to it)